### PR TITLE
Fixup broken case submission

### PR DESCRIPTION
**Pull request by nilueps**
The structure of cases.go was/is a mess. Hopefully the logic is a bit cleaner now. The behaviour of the api is also more obvious, with POST requests returning JSON validation objects when errors are encountered. The frontend still needs to be updated to handle this data appropriately in the way that the `individual.gohtml` template currently does. As an aside, I believe it might be worthwhile to move some of the validation logic to the client side, at least as far as syntax checking is concerned (emails, phone numbers, etc). Given that Core is intended to function in environments with spotty connection, we should reserve server-side validation for evaluations which are impossible to do on the client host.

### CODE REVIEW

File: api/pkg/apps/webapp/cases.go [hunk 1]
```diff
@@ -94,151 +94,63 @@ func (s *Server) Cases(w http.ResponseWriter, req *http.Request) {
...
-		team = teamRes
-	}
-
-	// Get parent case
-	if len(kase.ParentID) > 0 {
-		parent, err = cmsClient.Cases().Get(ctx, kase.ParentID)
-		if err != nil {
+	if !ok || len(caseID) == 0 {
+		if req.Method == "POST" {
+			caseID = "new"
```
Comment (keth) [inline api/pkg/apps/webapp/cases.go:121]: we probably want to use Constant for "new"

File: api/pkg/apps/webapp/cases.go [hunk 2]
```diff
@@ -247,8 +159,112 @@ func (s *Server) Case(w http.ResponseWriter, req *http.Request) {
 				s.Error(w, err)
 			}
 			return
+		}
+		var action string
+		if caseID == "new" {
+			action = "created"
```
Comment (keth) [inline api/pkg/apps/webapp/cases.go:165]: Missing Constants for all action types

File: api/pkg/apps/webapp/cases.go [hunk 3]
```diff
@@ -363,39 +377,35 @@ func (s *Server) NewCase(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
-func (s *Server) PutOrPostCase(req *http.Request, w http.ResponseWriter, ctx context.Context, kase *cms.Case) error {
+func (s *Server) PostCase(req *http.Request, w http.ResponseWriter, ctx context.Context, kase *cms.Case) (*cms.Case, error) {
```
Comment (keth) [inline api/pkg/apps/webapp/cases.go:380]: case with "k" ??  is this typo?

File: api/pkg/apps/webapp/cases.go [hunk 4]
```diff
@@ -264,9 +280,7 @@ func (s *Server) Case(w http.ResponseWriter, req *http.Request) {
 		"Comments":         displayComments(comments, commentAuthorMap),
 	}); err != nil {
 		s.Error(w, err)
-		return
 	}
-
 }
```
Comment (keth) [inline api/pkg/apps/webapp/cases.go:284]: This method is more than 150 lines of code.  I think we can chop it up a little, let's take a pair programming session together.

File: api/pkg/apps/webapp/cases.go (hunk 3 above)
Comment (nilueps) [reply api/pkg/apps/webapp/cases.go:380]: hehehe I guess you might also have been pulling your hair out trying to figure out why you were getting really weird syntax errors until you remember that `case` is a keyword :P

File: api/pkg/sessionmanager/interface.go [hunk 5]
```diff
@@ -85,33 +85,37 @@ const (
...
 func (r RedisSessionManager) AddNotification(req *http.Request, w http.ResponseWriter, notification *Notification) error {
-	session, err := r.Store.Get(req, varNotifications)
-	if err != nil {
-		return err
-	}
-	session.AddFlash(notification)
-	if err := session.Save(req, w); err != nil {
-		return err
-	}
+	// FIXME was converted to noop because of securecookie error
```
Comment (shinroo) [inline api/pkg/sessionmanager/interface.go:88]: Have we created tickets for these FIXMEs in JIRA?

Comment (nilueps) [reply api/pkg/sessionmanager/interface.go:88]: Now we have :)

**Review by keth — APPROVED** [review]

**Pull request merged**